Modify rule S2995: LaYC format (#2347)
SonarSource/rspec: rules/S2995/csharp/rule.adoc

== Why is this an issue?

> In C#, the https://learn.microsoft.com/en-us/dotnet/api/system.object.referenceequals[`Object.ReferenceEquals`] method is used to compare two https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/reference-types[reference type] variables. However, this method should not be used to compare https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/value-types[value types], such as `int`, `float`, or `bool`. This is because value types variable contains an instance of the type.
> 

Reviewer: ```suggestion
In C#, the https://learn.microsoft.com/en-us/dotnet/api/system.object.referenceequals[`Object.ReferenceEquals`] method is used to compare two https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/reference-types[reference type] variables. If you use this method to compare two https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/value-types[value types], such as `int`, `float`, or `bool` you will not get the expected results because value type variables contain an instance of the type and not a reference to it.
```

> Thus, value types variables can't have the same reference and using `Object.ReferenceEquals` to compare them will always return `false` even if the compared variables have the same value.

Reviewer: ```suggestion
Due to value type variables containing directly an instance of the type, they can't have the same reference, and using `Object.ReferenceEquals` to compare them will always return `false` even if the compared variables have the same value.
```

== How to fix it

=== Code examples

==== Noncompliant code example

[source,csharp,diff-id=1,diff-type=noncompliant]
----
public void Method(MyStruct struct1, MyStruct struct2)
{

>     if (Object.ReferenceEquals(struct1, struct2)) // Noncompliant: this will be always false

Reviewer: ```suggestion
    if (Object.ReferenceEquals(struct1, struct2)) // Noncompliant: this is always false
```

    {
      // ...
    }
}
----

==== Compliant solution

[source,csharp,diff-id=1,diff-type=compliant]
----
public void Method(MyStruct struct1, MyStruct struct2)
{

>     if (struct1.Equals(struct2)) // Compliant: value are compared

Reviewer: If I remember correctly in case it's not a default struct one should define the equality by overloading the comparison operators.

Might be better to give a more complete example in this case + add a sentence in "how to fix it?" WDTY?

See here: https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/how-to-define-value-equality-for-a-type#struct-example

Also please add this link to the resources - I think it's pretty useful.

Author: I updated the code examples to be more clear regarding value equality for structs.
